src/algorithms/union_find.hpp

```diff
@@ -0,0 +1,28 @@
+#pragma once
+
+#include <vector>
+
+struct UnionFind {
+	int new_var() {
+		int result = repr.size();
+		repr.push_back(result);
+		return result;
+	}
+
+	int find(int i) {
+		if (repr[i] == i)
+			return i;
+		return repr[i] = find(repr[i]);
+	}
+
+	void join_left_to_right(int i, int j) {
+		repr[find(i)] = find(j);
+	}
```

Comment:
I understand why we don't want to forget the direction of join, mainly because where we join stuff there's a "predominant" object (like a builtin function). In other words we need to choose the representative each time we join. Still, I'm not very happy with that long of a name in a well known algorithm that just calls this "join". I'm willing to keep it tho for the reason written above. Maybe "join_right" suffices?

Reply:
I kinda like this one tbh